map_creator/src/find_base_byAverage.cpp

```diff
@@ -0,0 +1,358 @@
+#include <octomap/math/Utils.h>
+
+#include <map_creator/sphere_discretization.h>
+#include <map_creator/kinematics.h>
+#include <map_creator/hdf5_dataset.h>
+#include <map_creator/WorkSpace.h>
+#include <map>
+
+#include <tf2/LinearMath/Transform.h>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2/LinearMath/Vector3.h>
+#include "geometry_msgs/PoseArray.h"
+
+#include <sys/types.h> 
+#include <sys/stat.h> 
+#include <ctime>
+
+#include "H5Cpp.h"
+#include <hdf5.h>
+#include <string>
+#include <time.h>
+
+using namespace H5;
+using namespace octomap;
+using namespace std;
+using namespace octomath;
+using namespace sphere_discretization;
+using namespace kinematics;
+using namespace hdf5_dataset;
+struct stat st;
+
+bool isIkSuccesswithTransformedBase(const geometry_msgs::Pose base_pose, const geometry_msgs::Pose grasp_pose, int& numOfSolns)
+{
+//Creating a transformation out of base pose
+    tf2::Vector3 base_vec(base_pose.position.x, base_pose.position.y, base_pose.position.z);
+    tf2::Quaternion base_quat(base_pose.orientation.x, base_pose.orientation.y, base_pose.orientation.z, base_pose.orientation.w);
+    base_quat.normalize();
+    tf2::Transform base_trns;
+    base_trns.setOrigin(base_vec);
+    base_trns.setRotation(base_quat);
+
+//Inverse of the transformation
+    tf2::Transform base_trns_inv;
+    base_trns_inv = base_trns.inverse();
+
+//Creating a transformation of grasp pose
+    tf2::Vector3 grasp_vec(grasp_pose.position.x, grasp_pose.position.y, grasp_pose.position.z);
+    tf2::Quaternion grasp_quat(grasp_pose.orientation.x, grasp_pose.orientation.y, grasp_pose.orientation.z, grasp_pose.orientation.w);
+    grasp_quat.normalize();
+    tf2::Transform grasp_trns;
+    grasp_trns.setOrigin(grasp_vec);
+    grasp_trns.setRotation(grasp_quat);
+
+//Transforming grasp pose to origin from where we can check for Ik
+    tf2::Transform new_grasp_trns;
+    //new_grasp_trns = grasp_trns * base_trns_inv;
+    new_grasp_trns =  base_trns_inv * grasp_trns;
+//Creating a new grasp pose in the origin co-ordinate
+    vector<double> new_grasp_pos;
+    tf2::Vector3 new_grasp_vec;
+    tf2::Quaternion new_grasp_quat;
+    new_grasp_vec =  new_grasp_trns.getOrigin();
+    new_grasp_quat =  new_grasp_trns.getRotation();
+    new_grasp_quat.normalize();
+    new_grasp_pos.push_back(new_grasp_vec[0]);
+    new_grasp_pos.push_back(new_grasp_vec[1]);
+    new_grasp_pos.push_back(new_grasp_vec[2]);
+    new_grasp_pos.push_back(new_grasp_quat[0]);
+    new_grasp_pos.push_back(new_grasp_quat[1]);
+    new_grasp_pos.push_back(new_grasp_quat[2]);
+    new_grasp_pos.push_back(new_grasp_quat[3]);
+
+//Check the new grasp_pose for Ik
+    Kinematics k;
+    std::vector<double> joints;
+    
+    joints.resize(6);
+    if(k.isIKSuccess(new_grasp_pos, joints, numOfSolns))
+      return true;
+    else
+      return false;
+
+} 
+
+bool areQuaternionClose(tf2::Quaternion q1, tf2::Quaternion q2)
+{
+     double dot = q1.dot(q2);
+     if(dot<0)
```

Comment:
Looking into the dot product of two quaternions, it looks to me like you might want a better threshold than 0.  I think that 0 means that the two quaternions are 90 degrees apart.  1 means that the quaternions are exactly aligned (zero angle between the pointing vectors).  So I would think that a value of 0.5-0.75 might be more appropriate.